### PR TITLE
Improvement + Fix: Cookie Buff Timer

### PR DESCRIPTION
**Pull request by ItsEmpa**
## What
Stores and uses the time until cookie buff runs out instead of depending on tablist. Also changes the formatting of it to show in only 1 line.

<details>
<summary>Images</summary>

image
image

</details>

## Changelog Improvements
+ Made Cookie Buff timer in Custom Scoreboard work without tab list. - Empa

## Changelog Fixes
+ Fixed bits available being subtracted in an infinite loop. - Empa

## Changelog Technical Details
+ Added Booster Cookie expiry time and hasCookieBuff to BitsAPI. - Empa
+ Added SimpleTimeMark.isInFuture(). - Empa

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 1]
```diff
@@ -39,6 +42,12 @@ object BitsAPI {
             profileStorage?.bitsToClaim = value
         }
 
+    var cookieBuffTime: Long
+        get() = profileStorage?.boosterCookieExpiryTime ?: 0
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
        get() = profileStorage?.boosterCookieExpiryTime.asTimeMark() ?: SimpleTimeMark.farPast()
```

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 2]
```diff
@@ -39,6 +42,12 @@ object BitsAPI {
...
+        private set(value) {
+            profileStorage?.boosterCookieExpiryTime = value
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
            profileStorage?.boosterCookieExpiryTime = value.toMillis()
````

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 3]
```diff
@@ -39,6 +42,12 @@ object BitsAPI {
             profileStorage?.bitsToClaim = value
         }
 
+    var cookieBuffTime: Long
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
    var cookieBuffTime: SimpleTimeMark
```

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 4]
```diff
@@ -156,22 +182,27 @@ object BitsAPI {
             // If the cookie stack is null, then the player should not have any bits to claim
             if (cookieStack == null) {
                 bitsToClaim = 0
+                cookieBuffTime = 0L
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
                cookieBuffTime = SimpleTimeMark.farPast()
```

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 5]
```diff
@@ -156,22 +182,27 @@ object BitsAPI {
...
                 bitsAvailableMenuPattern.matchMatcher(line) {
                     bitsToClaim = group("toClaim").formatInt()
-
-                    return
                 }
+                cookieDurationPattern.matchMatcher(line) {
+                    val duration = TimeUtils.getDuration(group("time"))
+                    cookieBuffTime = SimpleTimeMark.now().plus(duration).toMillis()
+                }
+                if (noCookieActiveSBMenuPattern.matches(line)) cookieBuffTime = 0L
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
                if (noCookieActiveSBMenuPattern.matches(line)) cookieBuffTime = SimpleTimeMark.farPast()
```

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 6]
```diff
@@ -212,6 +243,17 @@ object BitsAPI {
...
             }
+
+            line@ for (line in cookieStack.getLore()) {
+                cookieDurationPattern.matchMatcher(line) {
+                    val duration = TimeUtils.getDuration(group("time"))
+                    cookieBuffTime = SimpleTimeMark.now().plus(duration).toMillis()
+                }
+                if (noCookieActiveCookieMenuPattern.matches(line)) {
+                    val nextLine = cookieStack.getLore().nextAfter(line) ?: continue@line
+                    if (noCookieActiveCookieMenuPattern.matches(nextLine)) cookieBuffTime = 0L
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
                    if (noCookieActiveCookieMenuPattern.matches(nextLine)) cookieBuffTime = SimpleTimeMark.farPast()
```

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 7]
```diff
@@ -212,6 +243,17 @@ object BitsAPI {
                     continue@line
                 }
             }
+
+            line@ for (line in cookieStack.getLore()) {
+                cookieDurationPattern.matchMatcher(line) {
+                    val duration = TimeUtils.getDuration(group("time"))
+                    cookieBuffTime = SimpleTimeMark.now().plus(duration).toMillis()
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
                    cookieBuffTime = SimpleTimeMark.now().plus(duration)
```

File: src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt [hunk 8]
```diff
@@ -156,22 +182,27 @@ object BitsAPI {
...
 
             for (line in cookieStack.getLore()) {
                 bitsAvailableMenuPattern.matchMatcher(line) {
                     bitsToClaim = group("toClaim").formatInt()
-
-                    return
                 }
+                cookieDurationPattern.matchMatcher(line) {
+                    val duration = TimeUtils.getDuration(group("time"))
+                    cookieBuffTime = SimpleTimeMark.now().plus(duration).toMillis()
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/data/BitsAPI.kt]:
```suggestion
                    cookieBuffTime = SimpleTimeMark.now().plus(duration)
```

**Review by j10a1n15 — CHANGES_REQUESTED** [review]
Not tested when eating another cookie, rest works fine ingame
You could also add that image of the new design you send me

**Review by j10a1n15 — APPROVED** [review]
tested ingame (ate a cookie), worked fine

**Review by hannibal002 — CHANGES_REQUESTED** [review]
After first time updating to this, you need to open the SkyBlock menu once for the data to get loaded.
Please either show the red info "open sb menu" in the tab list or keep using the not precise tab list information until precise SkyBlock menu information is loaded.

File: src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ScoreboardElements.kt [hunk 9]
```diff
@@ -589,41 +588,18 @@ private fun getTuningDisplayPair(): List<Pair<String, HorizontalAlignment>> {
...
+        ?: "§cOpen SbMenu!") to HorizontalAlignment.LEFT
+)
 
 private fun getCookieShowWhen(): Boolean {
     if (HypixelData.bingo) return false
-
-    return if (informationFilteringConfig.hideEmptyLines) {
-        getCookieTime() != null
-    } else true
+    BitsAPI.cookieBuffTime?.let {
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ScoreboardElements.kt]:
```suggestion
    return BitsAPI.cookieBuffTime?.let {
        informationFilteringConfig.hideEmptyLines && BitsAPI.hasCookieBuff()
    } ?: true
```

**Review by j10a1n15 — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ScoreboardElements.kt (hunk 9 above)
Comment (j10a1n15) [reply src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ScoreboardElements.kt]: Whoops missclicked approved and request changes, that suggestion didnt seem to apply for all three lines too

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**